### ISSUE TITLE
AutoIncrement feature in MongoDB Attribute in Sails

### ISSUE THREAD

**Issue by rasekar2000** [balderdashy/sails#2641]
Hello Team,

 Will autoIncrement attribute in Sails ORM Waterline model for MongoDB?
 Please find more details here.
 http://stackoverflow.com/questions/28442345/sequence-number-for-mongodb-collection-in-sails-waterline

Thanks and Regards,
Raj

**Comment by MBrouns** [comment]
By default MongoDB does not have an option like auto increment. Instead it uses the objectID and stores it in _id. What is your specific use case that you require an auto increment?

**Comment by rasekar2000** [comment]
Hi @MBrouns  ,

Excuse me my RDBMS analogy/ usage in this case. The simple use case is that to store an employee in MongoDB via Waterline ( blueprint APIs ).  Let the _id be continue to be Object_ID ( yes Unique Index / Primary Key ) for Mongo DB. But the employee_number should be whole number , lets say 100 to 1000 and more..  Please note that the application will be deployed on the MongoDB  Shard Environment ( Multiple shardes and each consists of their own Replica Sets ) .  

   So, is that $inc the only option to address such scenario as per one of the suggestion section here... 
http://stackoverflow.com/questions/8384029/auto-increment-in-mongodb-to-store-sequence-of-unique-user-id

  I have seen that autoIncrement : true is an attribute available in Waterline Models. 
 https://github.com/balderdashy/waterline-schema.  I guess this 'autoIncrement : true' may be used in MySQL but not in MongoDB Adapter ..  So, should we consider $inc option as a solution to MongoDB for autoIncrement : true attribute? 

 Of should we need to rely on the Node JS like application to generate a auto increment value?

 Please suggest.

Thanks and Regards,
Raj

**Comment by MBrouns** [comment]
Well i'd say there's a fundamental issue when you're trying to have an auto-increment if you're gonna have multiple shards. If all of these shards are relatively autonomous and accept writes you're not gonna have a single authority to determine the sequence numbers. 

I would expect that this is also the reason that setting autoIncrement: true in Sails does not do anything when you're using MongoDB. It simply doesn't make a lot of sense to do so. 

There are a number of ways to get an auto incrementing field in Mongo (http://docs.mongodb.org/manual/tutorial/create-an-auto-incrementing-field/) but you should take a look at your use case and determine whether you actually need it. If so, i would suggest to pick one of the methods described in the link and implement that directly in your blueprints.

**Comment by rasekar2000** [comment]
Hi Mbrons,

  Yes.. I started this question from this article only .. (as per the initial question in Stackoverflow ) http://docs.mongodb.org/manual/tutorial/create-an-auto-incrementing-field/ 

 I think  optimistic lock is not a valid solution as the insert needs to be constantly validated for each Shard.

 But not sure about how Counters Collection loop will work as well in the shard environment. 

 Lets assume that the EMPLOYEE data is stored in few Shard Environments, where the Shard Keys are - Employee Id & Employee Number .. Lets Assume that the Employee_Id is actually the MongoDB's Object_Id, taken care by MongoDB engine. For the Employee Number - a nice whole number 1000 or 10000 or some valid whole number, it follows this  "Counters Collection" algorithm. 

 Under this scenario, my assumption is that insert statement will hit one of the many shard environment and the $inc is local to that shard environment and hence the resulting Order Number could go wrong. In other words the same Order number may be present in another Shard.. 
  Lets assume , we have mapped the Shard KEY ( Employee Number )  RANGE - ( 1 to 10000 ) for  Shared Instance 1 & 10001 - 20000 is for Shared Instance 2 and so on.. 

  Please note that I am OK to have few milliseconds delay in INSERT, if the MongoDB system needs to interact with other Sharded instance for this Counters Collection Loop. But I don't want to handle any Optimistic Loop like situation to handle the EXCEPTION scenario in the application.  

 So, in effective, I think Counters Collection may be used for my situation.. If not, I will have to keep the auto incrementing the number separately outside MongoDB. Thanks very much for your help.. 

Thanks and Regards,
Raj

**Comment by sailsbot** [comment]
Thanks for posting, @rasekar2000.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
